### PR TITLE
Generalize interceptors and add connection and transaction interception

### PR DESCRIPTION
**Pull request by ajcvickers**
Registration of interceptors is now command and moved to core:
```C#
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
    optionsBuilder
        .UseSqlite("DataSource=Test.db")
        .AddInterceptors(new MyRelationalInterceptor(), new MyLoggingInterceptor());
}

```

Creating chains (essentially multi-dispatch) is handled automatically, with a mechanism for new interceptors to integrate with the functionality.

Also, added a lot of async paths that will be needed for new ADO.NET async methods. Part of #15917

### CODE REVIEW

File: src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        ///     is to return the <paramref name="result" /> value passed in, often using <see cref="Task.FromResult{TResult}" />
+        /// </returns>
+        Task<InterceptionResult?> ConnectionClosingAsync(
+            [NotNull] DbConnection connection,
+            [NotNull] ConnectionEventData eventData,
+            InterceptionResult? result,
+            CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///     Called just after EF has called <see cref="DbConnection.Close()" /> in an async context.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs]: Not async

File: src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        /// </summary>
+        /// <param name="connection"> The connection. </param>
+        /// <param name="eventData"> Contextual information about the connection. </param>
+        void ConnectionClosed(
+            [NotNull] DbConnection connection,
+            [NotNull] ConnectionEndEventData eventData);
+
+        /// <summary>
+        ///     Called just after EF has called <see cref="DbConnection.Close()" />.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs]: Async

File: src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs]: Yes. These don't exist yet, until they are merged into .NET Standard 2.1. So this is just setting up for that.

File: src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        ///     Called immediately after EF calls <see cref="DbTransaction.Commit" />.
+        /// </summary>
+        /// <param name="transaction"> The transaction. </param>
+        /// <param name="eventData"> Contextual information about connection and transaction. </param>
+        void TransactionCommitted(
+            [NotNull] DbTransaction transaction,
+            [NotNull] TransactionEndEventData eventData);
+
+        /// <summary>
+        ///     Called just before EF intends to call <see cref="DbTransaction.Commit" />.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs]: Async

File: src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        ///     Called immediately after EF calls <see cref="DbTransaction.Rollback" />.
+        /// </summary>
+        /// <param name="transaction"> The transaction. </param>
+        /// <param name="eventData"> Contextual information about connection and transaction. </param>
+        void TransactionRolledBack(
+            [NotNull] DbTransaction transaction,
+            [NotNull] TransactionEndEventData eventData);
+
+        /// <summary>
+        ///     Called just before EF intends to call <see cref="DbTransaction.Rollback" />.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs]: Async

File: src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        ///     is to return the <paramref name="result" /> value passed in, often using <see cref="Task.FromResult{TResult}" />
+        /// </returns>
+        Task<InterceptionResult?> TransactionRollingBackAsync(
+            [NotNull] DbTransaction transaction,
+            [NotNull] TransactionEventData eventData,
+            InterceptionResult? result,
+            CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///     Called immediately after EF calls <see cref="DbTransaction.Rollback" />.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Diagnostics/IDbTransactionInterceptor.cs]: Async

**Pull request merged**